### ISSUE TITLE
FR: NewsDownloader feed text only

### ISSUE THREAD

**Issue by Frenzie** [koreader/koreader#3172]
* KOReader version: d4fd0b9
* Device: all

#### Issue
It doesn't always make sense to download the whole webpage.

#### Steps to reproduce
Use a feed with full text such as http://ftr.fivefilters.org/makefulltextfeed.php?url=http%3A%2F%2Fwww.nu.nl%2Frss%2FAlgemeen&max=3 or https://github.com/koreader/koreader/releases.atom

PS That first feed apparently doesn't work, see https://www.mobileread.com/forums/showthread.php?p=3576634#post3576634

PPS Pinging @mwoz123. :-)

Edit: I can't confirm any issues with http://ftr.fivefilters.org/makefulltextfeed.php?url=http%3A%2F%2Fwww.nu.nl%2Frss%2FAlgemeen&max=3 btw.

**Comment by mwoz123** [comment]
Looks very interesting. 

But basic, "free" service has a [limit of max 3](http://fivefilters.org/content-only/#section-pricing) news (feed) per url (item). IMHO this is too low.

Any other services with bigger or no limit?

**Comment by Frenzie** [comment]
The FR doesn't really have anything to do with that particular service. :-) (But incidentally, you can [host your own](https://bitbucket.org/fivefilters/full-text-rss) and there's also the neat [rss-bridge](https://github.com/RSS-Bridge/rss-bridge).)

**Comment by mwoz123** [comment]
Quick glance and they are both seems PHP applications - I probably won't be able to help.

This would be such big change that I'd suggest to create new plugin e.g. "NewsDownloader2" (from current NewsDownloader code). 
So unit it's fully finished & working we can still have working "NewsDownloader 1" in nightly (or both).

**Comment by Frenzie** [comment]
I think we have a slight misunderstanding. For a full-text feed, such as https://github.com/koreader/koreader/releases.atom (forget about fivefilters, it's just another way to get one), it doesn't make sense to download the whole webpage.

Turning a partial feed into a full feed is an entirely different can of worms.

**Comment by mwoz123** [comment]
So how this enhancement should be done? What you want to use?
Would it have similar architecture:
Koreader N.D. -> sends feeds urls to a server-> server download feeds -> server does fulltext search-> server sends back data (text) -> to korader N.D.

**Comment by Frenzie** [comment]
I would start by renaming `downloadFeed` to something else because that's not what it does and it seems to be fueling a lot of confusion. `downloadPage` might be a better name.

https://github.com/koreader/koreader/blob/74242025ff7734006a04f016d9942271ed6fc86c/plugins/newsdownloader.koplugin/main.lua#L262

With that problem tackled we can save a simple HTML file.
We have `feed.title` and we have `feed.item.content`.

```
<!DOCUMENT html>
<title>feed.title</title>
feed.item.content
```

That's pretty much all you need. I don't know why you want 50 server round trips. ;-)

**Comment by mwoz123** [comment]
@Frenzie this seems completely different that what I guess was initially proposed.

Most of my feeds in `feed.item.context` provide only:
-  first paragraph with 
- link "to read rest of the article click here". 

That's why I use `feed.url` and download full page instead of doing what you proposed in you last comment...

**Comment by mwoz123** [comment]
So I'd get back to initial idea..

**Comment by Frenzie** [comment]
No, I am and always have asked for feed text only. No server round trips.

> Use a feed with full text such as http://ftr.fivefilters.org/makefulltextfeed.php?url=http%3A%2F%2Fwww.nu.nl%2Frss%2FAlgemeen&max=3 or https://github.com/koreader/koreader/releases.atom

You already have the full text, but the program goes to download each and every single linked page. This wastes time and gives you an inferior end result.

https://github.com/koreader/koreader/issues/3172#issuecomment-327473633
> I think we have a slight misunderstanding. For a full-text feed, such as https://github.com/koreader/koreader/releases.atom (forget about fivefilters, it's just another way to get one), it doesn't make sense to download the whole webpage.

**Comment by Frenzie** [comment]
But that aside, the whole text downloading (which should not be the default imo) would be hugely improved by some simple className filtering.

**Comment by mwoz123** [comment]
Let's first start with configuration UI. 
Then let's add such feature. 
It'd be easier to change it from (your)  'default' to (my)  'current' behavior. 
And as I said nowadays many (most ?) of the RSS in item context holds only couple sentences +link to web page for 'read more'. 
SoI think it might be needed often than you expect..

**Comment by mwoz123** [comment]
So maybe you were right @Fenzie that we should also have option (we've to discuss which behaviour should be default) to use `item.context` as data source. I see some of the feed.item.url returns http 302 status (redirect) which is not handled by this code:(
But migration configuration will require availability to save our feed_config.lua

**Comment by Frenzie** [comment]
Are these https feeds?

**Comment by mwoz123** [comment]
No.

**Comment by Frenzie** [comment]
Redirects don't work on luasec — a reason I've been toying with the idea of switching to lua-http — but they should on luasocket. Have you tried debugging it? What do you mean by "not handled"?

**Comment by Thra11** [comment]
Since the epub-with-images changes (#4894), with `download_full_article = false`, it still writes an html file, but it names it .epub. KOReader doesn't seem to care (it still opens it), but it should probably be fixed. Which raises the question of whether we want to put the html content from the feed through the same process as the full article html, to produce an epub with (optional) images included. I think you can have images in the content of an rss/atom feed, can't you?

**Comment by Thra11** [comment]
It should be simple enough to lift the `local html = self:loadPage(url)` out of `EpubDownloadBackend:createEpub`, and change `createEpub` to take the html instead of the url (actually it would probably need to take both, so that it can still deduce the base_url to handle relative urls).

**Comment by Frenzie** [comment]
> I think you can have images in the content of an rss/atom feed, can't you?

Of course! ;-)

**Comment by Thra11** [comment]
> whole text downloading (which should not be the default imo)

I'm inclined to agree. When the feed contains the full article content, using that produces better results than downloading the url. And if the feed only contains a summary or just the link with no content at all, then the footer text which gets added lets the user know how to switch to downloading full articles for that feed (whereas there's no equivalent, "If this article was full of headers and menus, you might want to try turning off download_full_article", and if there was it would be harder to find in the chaos).

I'd be curious to know what percentage of feeds include the full text versus a summary versus no content, just links, as I'm currently basing my opinion on a sample of one rss feed and one atom feed.

**Comment by Frenzie** [comment]
> I'd be curious to know what percentage of feeds include the full text versus a summary versus no content, just links, as I'm currently basing my opinion on a sample of one rss feed and one atom feed.

It probably depends on the type. I figure non-commercial is mostly complete, while fast-moving and/or commercial is mostly summaries.

**Comment by mwoz123** [comment]
**TLDR**: each saved news should contain only main article text (no nav bar, footer, possibly ads, external links etc). 
There is a service (or rather code in PHP- fivefilters.org) that could do that server side - imo that's the main idea here

**Comment by Frenzie** [comment]
The tl;dr is the OP. A full-text feed such as [please ignore the FiveFilters example] **or https://github.com/koreader/koreader/releases.atom**.

There is no "main article text". There is just feed text.

**Comment by Thra11** [comment]
What remains to be done for this issue? As far as I'm aware, setting `download_full_article = false` will use the text from the feed (optionally fetching and including images if `include_images` is true). Did you have any other specific improvements in mind before this issue can be closed?

**Comment by Frenzie** [comment]
I still think the default is quite bad but I don't know why this isn't closed.

**Comment by georgew21** [comment]
hello, I also think that it's problem that this plugin downloads full webpage. it could be far more useuful if it was possible to download only the text from the article.

 also, it could wonderful if it was downloading all articles of a feed in a single epub with contents.

i think that the fetch news from calibre should be the ideal function ,but i don' know if this is possible in koreader.

**Comment by Frenzie** [comment]
It can be changed with `download_full_article = false`. I'll close this now.

**Comment by georgew21** [comment]
but "download_full_article = false" downloads only the description of the article and not the full text of article. on the other side, download_full_article = true downloads full webpage which is unconvient because the article can be in last pages of html.

 the solution should be like fetch news in calibre which keep only specific tags, like body etc

**Comment by Frenzie** [comment]
This FR was about not making extra round trips; feel free to open another one. ;-)

**Comment by Frenzie** [comment]
My bad, indeed the plugin does *not* do what this FR asks for.

**Comment by Frenzie** [comment]
Might be as simple as:

```diff
diff --git a/plugins/newsdownloader.koplugin/main.lua b/plugins/newsdownloader.koplugin/main.lua
index a5d3ec47..a1093c73 100644
--- a/plugins/newsdownloader.koplugin/main.lua
+++ b/plugins/newsdownloader.koplugin/main.lua
@@ -318,7 +318,7 @@ function NewsDownloader:processRSS(feeds, limit, download_full_article, include_
         if download_full_article then
             self:downloadFeed(feed, feed_output_dir, include_images, article_message, enable_filter, filter_element)
         else
-            self:createFromDescription(feed, feed.description, feed_output_dir, include_images, article_message)
+            self:createFromDescription(feed, feed["content:encoded"] or feed.description, feed_output_dir, include_images, article_message)
         end
     end
 end
```
